google/cloud/storage/internal/rest_client.cc

```diff
@@ -53,6 +54,56 @@ StatusOr<ReturnType> ParseFromRestResponse(
   return ReturnType::FromHttpResponse(*payload);
 }
 
+template <typename Parser>
+auto CheckedFromString(StatusOr<std::unique_ptr<rest::RestResponse>> response)
+    -> decltype(Parser::FromString(
+        *(rest::ReadAll(std::move(**response).ExtractPayload())))) {
+  if (!response.ok()) {
+    return std::move(response).status();
+  }
+  if ((*response)->StatusCode() >= rest::kMinNotSuccess) {
+    return rest::AsStatus(std::move(**response));
+  }
+  auto payload = rest::ReadAll(std::move(**response).ExtractPayload());
+  if (!payload.ok()) return std::move(payload).status();
+  return Parser::FromString(*payload);
+}
+
+StatusOr<EmptyResponse> ReturnEmptyResponse(
+    StatusOr<std::unique_ptr<rest::RestResponse>> response) {
+  if (!response.ok()) {
+    return std::move(response).status();
+  }
+  if ((*response)->StatusCode() >= rest::kMinNotSuccess) {
+    return rest::AsStatus(std::move(**response));
+  }
+  return EmptyResponse{};
+}
+
+template <typename ReturnType>
+StatusOr<ReturnType> CreateFromJson(
+    StatusOr<std::unique_ptr<rest::RestResponse>> response) {
+  if (!response.ok()) return std::move(response).status();
+  if ((*response)->StatusCode() >= rest::kMinNotSuccess) {
+    return rest::AsStatus(std::move(**response));
+  }
+  auto payload = rest::ReadAll(std::move(**response).ExtractPayload());
+  if (!payload.ok()) return std::move(payload).status();
+  return ReturnType::CreateFromJson(*payload);
+}
+
+Status AddAuthorizationHeader(Options const& options,
+                              RestRequestBuilder& builder) {
+  auto auth_header =
+      options.get<Oauth2CredentialsOption>()->AuthorizationHeader();
+  if (!auth_header.ok()) {
+    return std::move(auth_header).status();
+  }
+  builder.AddHeader("Authorization", std::string(absl::StripPrefix(
+                                         *auth_header, "Authorization: ")));
+  return {};
+}
+
```

Comment:
On a future PR (maybe): this class would be unit testable if the constructor accepted `storage_rest_client_`, `iam_rest_client_`, and `curl_client_` as parameters.